File: code/programs/typescript/unix-tools/src/realpath.ts
/**
 * realpath -- print the resolved absolute file name.
 *
 * === What This Program Does ===
 *
 * This is a reimplementation of the GNU `realpath` utility in TypeScript.
 * For each FILE, it prints the resolved absolute pathname. All symbolic
 * links, `.` and `..` components are resolved.
 *
 * === How realpath Works ===
 *
 * realpath resolves a path to its canonical form:
 *
 *     realpath .                    =>   /home/user/projects
 *     realpath ../foo               =>   /home/user/foo
 *     realpath /usr/bin/../lib      =>   /usr/lib
 *     realpath symlink              =>   /actual/target/path
 *
 * === Canonicalization Modes ===
 *
 * By default (no flags), all path components must exist and symlinks are
 * resolved. The flags modify this behavior:
 *
 * - `-e` (canonicalize-existing): All components must exist (strict).
 * - `-m` (canonicalize-missing): No component needs to exist (lenient).
 * - `-s` (no-symlinks): Don't resolve symlinks, just normalize the path.
 *
 * === Relative Output ===
 *
 * `--relative-to=DIR` prints the result relative to DIR.
 * `--relative-base=DIR` prints relative if the path starts with DIR,
 * otherwise prints the absolute path.
 *
 * === Zero-Terminated Output (-z) ===
 *
 * With `-z`, each output line ends with NUL instead of newline. This is
 * useful for piping to `xargs -0`.
 *
 * @module realpath
 */

import * as fs from "node:fs";
import * as path from "node:path";
import { fileURLToPath } from "node:url";

// ---------------------------------------------------------------------------
// Import CLI Builder.
// ---------------------------------------------------------------------------

import { Parser } from "@coding-adventures/cli-builder";

// ---------------------------------------------------------------------------
// Locate the JSON spec file.
// ---------------------------------------------------------------------------

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);
const SPEC_FILE = path.resolve(__dirname, "..", "realpath.json");

// ---------------------------------------------------------------------------
// Business Logic: Resolve paths.
// ---------------------------------------------------------------------------

/**
 * Resolve a path to its canonical form.
 *
 * The behavior depends on the mode:
 * - Default / `-e`: Use fs.realpathSync, which requires all components to
 *   exist and resolves symlinks.
 * - `-m`: Resolve what we can, normalize the rest. No component needs to exist.
 * - `-s`: Don't resolve symlinks, just normalize with path.resolve.
 */
function resolvePath(
  filePath: string,
  canonicalizeExisting: boolean,
  canonicalizeMissing: boolean,
  noSymlinks: boolean
): string {
  if (noSymlinks) {
    // Just normalize the path without resolving symlinks.
    return path.resolve(filePath);
  }

  if (canonicalizeMissing) {
    // Resolve as much as possible. For missing components, just normalize.
    // We try realpathSync first; if it fails, fall back to path.resolve.
    try {
      return fs.realpathSync(filePath);
    } catch {
      return path.resolve(filePath);
    }
  }

  // Default and -e: all components must exist.
  // fs.realpathSync throws if any component is missing.
  return fs.realpathSync(filePath);
}

// ---------------------------------------------------------------------------
// Main: parse args via CLI Builder, then resolve paths.
// ---------------------------------------------------------------------------

function main(): void {
  // --- Step 1: Parse arguments ---------------------------------------------

  let result;

  try {
    const parser = new Parser(SPEC_FILE, process.argv);
    result = parser.parse();
  } catch (err: unknown) {
    if (err && typeof err === "object" && "errors" in err) {
      const errors = (err as { errors: Array<{ message: string }> }).errors;
      for (const error of errors) {
        process.stderr.write(`realpath: ${error.message}\n`);
      }
      process.exit(1);
    }
    throw err;
  }

  // --- Step 2: Dispatch on result type -------------------------------------

  if ("text" in result) {
    process.stdout.write(result.text + "\n");
    process.exit(0);
  }

  if ("version" in result && !("flags" in result)) {
    process.stdout.write(result.version + "\n");
    process.exit(0);
  }

  // --- Step 3: Extract flags and arguments ---------------------------------

  const flags = (result as { flags: Record<string, unknown> }).flags;
  const args = (result as { arguments: Record<string, unknown> }).arguments;

  const canonicalizeExisting = !!flags["canonicalize_existing"];
  const canonicalizeMissing = !!flags["canonicalize_missing"];
  const noSymlinks = !!flags["no_symlinks"];
  const quiet = !!flags["quiet"];
  const relativeTo = flags["relative_to"] as string | undefined;
  const relativeBase = flags["relative_base"] as string | undefined;
  const zero = !!flags["zero"];

  const delimiter = zero ? "\0" : "\n";

  // Normalize file list.
  let files = args["files"] as string[] | string;
  if (typeof files === "string") {
    files = [files];
  }

  // --- Step 4: Resolve and print each path ---------------------------------

  for (const file of files) {
    try {
      let resolved = resolvePath(
        file,
        canonicalizeExisting,
        canonicalizeMissing,
        noSymlinks
      );

      // Apply relative-to or relative-base if specified.
      if (relativeTo) {
        const resolvedRelTo = resolvePath(
          relativeTo,
          canonicalizeExisting,
          canonicalizeMissing,
          noSymlinks
        );
        resolved = path.relative(resolvedRelTo, resolved);
      } else if (relativeBase) {
        const resolvedBase = resolvePath(
          relativeBase,
          canonicalizeExisting,
          canonicalizeMissing,
          noSymlinks
        );
        if (resolved.startsWith(resolvedBase)) {
          resolved = path.relative(resolvedBase, resolved);
        }
        // Otherwise, keep the absolute path.
      }

      process.stdout.write(resolved + delimiter);
    } catch (err: unknown) {
      if (!quiet) {
        const message = err instanceof Error ? err.message : String(err);
        process.stderr.write(`realpath: ${file}: ${message}\n`);
      }
      process.exitCode = 1;
    }
  }
}

// ---------------------------------------------------------------------------
// Run the program.
// ---------------------------------------------------------------------------

main();

File: code/programs/elixir/unix-tools/lib/unix_tools/realpath.ex
defmodule UnixTools.Realpath do
  @moduledoc """
  realpath -- print the resolved absolute file name.

  ## What This Program Does

  This is a reimplementation of the GNU `realpath` utility in Elixir. For each
  FILE, it prints the resolved absolute pathname. All symbolic links, `.` and
  `..` components are resolved.

  ## How realpath Works

  realpath resolves a path to its canonical form:

      realpath .                    =>   /home/user/projects
      realpath ../foo               =>   /home/user/foo
      realpath /usr/bin/../lib      =>   /usr/lib

  ## Canonicalization Modes

  - Default: all components must exist, symlinks are resolved.
  - `-e`: All components must exist (strict, same as default).
  - `-m`: No component needs to exist (lenient).
  - `-s`: Don't resolve symlinks, just normalize.

  ## Relative Output

  `--relative-to=DIR` prints the result relative to DIR.
  `--relative-base=DIR` prints relative if the path starts with DIR.
  """

  alias CodingAdventures.CliBuilder.Parser
  alias CodingAdventures.CliBuilder.{ParseResult, HelpResult, VersionResult, ParseErrors}

  # ---------------------------------------------------------------------------
  # Entry point
  # ---------------------------------------------------------------------------

  @doc """
  Entry point. Receives `argv` as a list of strings.
  """
  def main(argv) do
    spec_path = resolve_spec_path()

    case Parser.parse(spec_path, ["realpath" | argv]) do
      {:ok, %HelpResult{text: text}} ->
        IO.puts(text)

      {:ok, %VersionResult{version: version}} ->
        IO.puts(version)

      {:ok, %ParseResult{flags: flags, arguments: arguments}} ->
        canonicalize_existing = !!flags["canonicalize_existing"]
        canonicalize_missing = !!flags["canonicalize_missing"]
        no_symlinks = !!flags["no_symlinks"]
        quiet = !!flags["quiet"]
        relative_to = flags["relative_to"]
        relative_base = flags["relative_base"]
        zero = !!flags["zero"]

        file_delimiter = if zero, do: <<0>>, else: "\n"
        file_list = normalize_files(arguments["files"])

        Enum.each(file_list, fn file ->
          resolve_and_print(
            file,
            canonicalize_existing,
            canonicalize_missing,
            no_symlinks,
            quiet,
            relative_to,
            relative_base,
            file_delimiter
          )
        end)

      {:error, %ParseErrors{errors: errors}} ->
        Enum.each(errors, fn e ->
          IO.puts(:stderr, "realpath: #{e.message}")
        end)

        System.halt(1)
    end
  end

  # ---------------------------------------------------------------------------
  # Business Logic: Path resolution.
  # ---------------------------------------------------------------------------

  @doc """
  Resolve a path to its canonical form.

  Uses different strategies based on the mode flags:
  - `-s`: Just normalize with `Path.expand/1` (no symlink resolution).
  - `-m`: Try to resolve, fall back to `Path.expand/1` for missing paths.
  - Default/`-e`: Resolve fully, error if any component is missing.
  """
  def resolve_path(file_path, _canonicalize_existing, _canonicalize_missing, true = _no_symlinks) do
    {:ok, Path.expand(file_path)}
  end

  def resolve_path(file_path, _canonicalize_existing, true = _canonicalize_missing, _no_symlinks) do
    # Try to resolve symlinks; fall back to expand for missing paths.
    case resolve_with_readlink(file_path) do
      {:ok, resolved} -> {:ok, resolved}
      {:error, _} -> {:ok, Path.expand(file_path)}
    end
  end

  def resolve_path(file_path, _canonicalize_existing, _canonicalize_missing, _no_symlinks) do
    # Default and -e: all components must exist.
    resolve_with_readlink(file_path)
  end

  defp resolve_with_readlink(file_path) do
    # Expand the path first to resolve . and .. components.
    expanded = Path.expand(file_path)

    # Check if the file exists (following symlinks).
    case File.stat(expanded) do
      {:ok, _stat} ->
        # File exists. Now resolve any symlinks in the path.
        case :file.read_link_all(to_charlist(expanded)) do
          {:ok, target} ->
            # It's a symlink -- resolve the target.
            target_path = to_string(target)
            abs_target = Path.expand(target_path, Path.dirname(expanded))
            {:ok, abs_target}

          {:error, :einval} ->
            # Not a symlink -- the expanded path is canonical.
            {:ok, expanded}

          {:error, _} ->
            {:ok, expanded}
        end

      {:error, reason} ->
        {:error, reason}
    end
  end

  defp resolve_and_print(
         file,
         canonicalize_existing,
         canonicalize_missing,
         no_symlinks,
         quiet,
         relative_to,
         relative_base,
         file_delimiter
       ) do
    case resolve_path(file, canonicalize_existing, canonicalize_missing, no_symlinks) do
      {:ok, resolved} ->
        # Apply relative-to or relative-base adjustments.
        final =
          cond do
            relative_to != nil ->
              case resolve_path(relative_to, canonicalize_existing, canonicalize_missing, no_symlinks) do
                {:ok, base} -> Path.relative_to(resolved, base)
                {:error, _} -> resolved
              end

            relative_base != nil ->
              case resolve_path(relative_base, canonicalize_existing, canonicalize_missing, no_symlinks) do
                {:ok, base} ->
                  if String.starts_with?(resolved, base) do
                    Path.relative_to(resolved, base)
                  else
                    resolved
                  end

                {:error, _} ->
                  resolved
              end

            true ->
              resolved
          end

        IO.write(final <> file_delimiter)

      {:error, reason} ->
        unless quiet do
          IO.puts(:stderr, "realpath: #{file}: #{:file.format_error(reason)}")
        end
    end
  end

  # ---------------------------------------------------------------------------
  # Helpers
  # ---------------------------------------------------------------------------

  defp normalize_files(files) when is_list(files), do: files
  defp normalize_files(file) when is_binary(file), do: [file]

  defp resolve_spec_path do
    candidates = [
      if(function_exported?(Mix, :Project, 0),
        do: Path.join(Mix.Project.config()[:lockfile] |> Path.dirname(), "realpath.json"),
        else: nil
      ),
      "realpath.json",
      Path.join(:code.priv_dir(:unix_tools) |> to_string(), "realpath.json")
    ]

    Enum.find(candidates, fn
      nil -> false
      candidate_path -> File.exists?(candidate_path)
    end) ||
      raise "Could not find realpath.json spec file"
  end
end
